README.md
[![htp crate](https://img.shields.io/crates/v/htp.svg)](https://crates.io/crates/htp)
[![tbl documentation](https://docs.rs/htp/badge.svg)](https://docs.rs/htp)
[![GitHub license](https://img.shields.io/github/license/PicoJr/htp)](https://github.com/PicoJr/htp/blob/master/LICENSE)
# HTP

Work in progress **H**uman **T**ime **P**arser

This lib uses [pest](https://github.com/pest-parser/pest) for parsing.

## Example

```rust
use chrono::{Utc, TimeZone};
use htp::parse;
let now = Utc.datetime_from_str("2020-12-24T23:45:00", "%Y-%m-%dT%H:%M:%S").unwrap();
let expected = Utc.datetime_from_str("2020-12-18T19:43:00", "%Y-%m-%dT%H:%M:%S").unwrap();
let datetime = parse("last friday at 19:43", now).unwrap();
assert_eq!(datetime, expected);
```

## Similar Crate

* [chrono-english](https://github.com/stevedonovan/chrono-english).

## Why?

Tweak how time is parsed and interpreted inside my [rust time tracking tool](https://github.com/PicoJr/rtw).

It's fun to write parsers once in while, [pest](https://github.com/pest-parser/pest) is really nice.

## What date format can it parse?

see [time_clue grammar rule](src/time.pest).

some examples:

* `4 min ago`, `4 h ago`, `1 week ago`, `in 2 hours`, `in 1 month`
* `last friday at 19`, `monday at 6 am`
* `7`, `7am`, `7pm`, `7:30`, `19:43:00`
* `now`, `yesterday`, `today`, `friday`
* `2020-12-25T19:43:00`

It also supports _interestingly-spaced_ inputs such as:
```
4           min      ago
```

It is possible to try `HTP` out using `cargo run --example time_parser`:

example
```
cargo run --example time_parser last friday at 6
```

output
```
2020-07-03T06:00:00+02:00
```

Thanks to pest it also provides meaningful errors:

example
```
cargo run --example time_parser last friday at
```

output
```
 --> 1:15
  |
1 | last friday at
  |               ^---
  |
  = expected hms
```

## Changelog

Please see the [CHANGELOG](CHANGELOG.md) for a release history.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.4.0](https://crates.io/crates/htp/0.4.0) Sep 18, 2021

* Add `tomorrow` ([#2](https://github.com/PicoJr/htp/pull/2) by [@geobert](https://github.com/Geobert))
* Add `parse_time_clue` and `evaluate_time_clue` ([#3](https://github.com/PicoJr/htp/pull/3) by [@geobert](https://github.com/Geobert))
* Add support for `/`-separated dates such as `2020/25/12` ([#4](https://github.com/PicoJr/htp/pull/4) by [@geobert](https://github.com/Geobert))

## [0.3.0](https://crates.io/crates/htp/0.3.0) Jul 7, 2021

* Add `in <quantifier>` time clue ([#1](https://github.com/PicoJr/htp/pull/1) by [@bmoelk](https://github.com/bmoelk))

## [0.2.1](https://crates.io/crates/htp/0.2.1) Jul 12, 2020

* Fix `next <weekday>` evaluation.
* Add test

## [0.2.0](https://crates.io/crates/htp/0.2.0) Jul 05, 2020

* Add am,pm
* Add weekdays and quantifiers short aliases

## [0.1.0](https://crates.io/crates/htp/0.1.0) Jul 04, 2020

* Add `parse`
* Add `examples`
* Add `README`, `LICENSE`, `CHANGELOG`


src/lib.rs
//! # HTP
//! (H)uman (T)ime (P)arser.
//!
//! ## Example
//!
//! ```
//! use chrono::{Utc, TimeZone};
//! use htp::parse;
//! let now = Utc.datetime_from_str("2020-12-24T23:45:00", "%Y-%m-%dT%H:%M:%S").unwrap();
//! let expected = Utc.datetime_from_str("2020-12-18T19:43:00", "%Y-%m-%dT%H:%M:%S").unwrap();
//! let datetime = parse("last friday at 19:43", now).unwrap();
//! assert_eq!(datetime, expected);
//! ```
//!
extern crate pest;
#[macro_use]
extern crate pest_derive;

use chrono::DateTime;
use thiserror::Error;

pub mod interpreter;
pub mod parser;

#[derive(Error, Debug)]
pub enum HTPError {
    #[error(transparent)]
    ParseError(#[from] parser::ParseError),
    #[error(transparent)]
    EvaluationError(#[from] interpreter::EvaluationError),
}

/// Same as `parse_time_clue(s, now, false)`
///
/// Parse time clue from `s` given reference time `now` in timezone `Tz`.
pub fn parse<Tz: chrono::TimeZone>(s: &str, now: DateTime<Tz>) -> Result<DateTime<Tz>, HTPError> {
    parse_time_clue(s, now, false)
}

/// Parse time clue from `s` given reference time `now` in timezone `Tz`.
///
/// `assume_next_day`:
/// * if true: times without a day will be interpreted as times during the following the day.
/// e.g. 19:43 will be interpreted as tomorrow at 19:43 if current time is > 19:43.
/// * if false: times without a day will be interpreted as times during current day.
pub fn parse_time_clue<Tz: chrono::TimeZone>(
    s: &str,
    now: DateTime<Tz>,
    assume_next_day: bool,
) -> Result<DateTime<Tz>, HTPError> {
    let time_clue = parser::parse_time_clue_from_str(s)?;
    let datetime = interpreter::evaluate_time_clue(time_clue, now, assume_next_day)?;
    Ok(datetime)
}


src/interpreter.rs
use crate::parser::{Modifier, Quantifier, ShortcutDay, TimeClue, AMPM, HMS};
use chrono::{DateTime, Datelike, Duration, LocalResult, TimeZone, Utc};
use thiserror::Error;

#[derive(Error, Debug, PartialEq)]
pub enum EvaluationError {
    #[error("invalid time: {hour}:{minute}:{second} {am_or_pm}")]
    InvalidTimeAMPM {
        hour: u32,
        minute: u32,
        second: u32,
        am_or_pm: AMPM,
    },
    #[error("invalid time: {hour}:{minute}:{second}")]
    InvalidTime { hour: u32, minute: u32, second: u32 },
    #[error("invalid ISO date: {year}-{month}-{day}T{hour}:{minute}:{second}")]
    ChronoISOError {
        year: i32,
        month: u32,
        day: u32,
        hour: u32,
        minute: u32,
        second: u32,
    },
}

fn check_hms(hms: HMS, am_or_pm_maybe: Option<AMPM>) -> Result<HMS, EvaluationError> {
    let (h, m, s) = hms;
    let h_am_pm = match am_or_pm_maybe {
        None | Some(AMPM::AM) => h,
        Some(AMPM::PM) => h + 12,
    };
    if h_am_pm < 24 && m < 60 && s < 60 {
        Ok((h_am_pm, m, s))
    } else {
        match am_or_pm_maybe {
            Some(am_or_pm) => Err(EvaluationError::InvalidTimeAMPM {
                hour: h,
                minute: m,
                second: s,
                am_or_pm,
            }),
            None => Err(EvaluationError::InvalidTime {
                hour: h,
                minute: m,
                second: s,
            }),
        }
    }
}

/// Same as `evaluate(time_clue, now)`
pub fn evaluate<Tz: chrono::TimeZone>(
    time_clue: TimeClue,
    now: DateTime<Tz>,
) -> Result<DateTime<Tz>, EvaluationError> {
    evaluate_time_clue(time_clue, now, false)
}

/// Evaluate `time_clue` given reference time `now`.
///
/// `assume_next_day`:
/// * if true: times without a day will be interpreted as times during the following the day.
/// e.g. 19:43 will be interpreted as tomorrow at 19:43 if current time is > 19:43.
/// * if false: times without a day will be interpreted as times during current day.
pub fn evaluate_time_clue<Tz: chrono::TimeZone>(
    time_clue: TimeClue,
    now: DateTime<Tz>,
    assume_next_day: bool, // assume next day if only time is supplied and time < now
) -> Result<DateTime<Tz>, EvaluationError> {
    match time_clue {
        TimeClue::Now => Ok(now),
        TimeClue::Time((h, m, s), am_or_pm_maybe) => {
            let (h, m, s) = check_hms((h, m, s), am_or_pm_maybe)?;
            let d = now.date().and_hms(h, m, s);
            if assume_next_day && d < now {
                Ok(d + Duration::days(1))
            } else {
                Ok(d)
            }
        }
        TimeClue::Relative(n, quantifier) => match quantifier {
            Quantifier::Min => Ok(now - Duration::minutes(n as i64)),
            Quantifier::Hours => Ok(now - Duration::hours(n as i64)),
            Quantifier::Days => Ok(now - Duration::days(n as i64)),
            Quantifier::Weeks => Ok(now - Duration::weeks(n as i64)),
            Quantifier::Months => Ok(now - Duration::days(30 * n as i64)), // assume 1 month = 30 days
        },
        TimeClue::RelativeFuture(n, quantifier) => match quantifier {
            Quantifier::Min => Ok(now + Duration::minutes(n as i64)),
            Quantifier::Hours => Ok(now + Duration::hours(n as i64)),
            Quantifier::Days => Ok(now + Duration::days(n as i64)),
            Quantifier::Weeks => Ok(now + Duration::weeks(n as i64)),
            Quantifier::Months => Ok(now + Duration::days(30 * n as i64)), // assume 1 month = 30 days
        },
        TimeClue::RelativeDayAt(modifier, weekday, hms_maybe, am_or_pm_maybe) => {
            let (h, m, s) = hms_maybe.unwrap_or((0, 0, 0));
            let (h, m, s) = check_hms((h, m, s), am_or_pm_maybe)?;
            let monday = now.date() - Duration::days(now.weekday().num_days_from_monday() as i64);
            match modifier {
                Modifier::Last => {
                    let same_week_day =
                        monday + (Duration::days(weekday.num_days_from_monday() as i64));
                    if weekday.num_days_from_monday() < now.weekday().num_days_from_monday() {
                        Ok(same_week_day.and_hms(h, m, s)) // same week
                    } else {
                        Ok(same_week_day.and_hms(h, m, s) - Duration::days(7)) // last week
                    }
                }
                Modifier::Next => {
                    let same_week_day =
                        monday + (Duration::days(weekday.num_days_from_monday() as i64));
                    if weekday.num_days_from_monday() > now.weekday().num_days_from_monday() {
                        Ok(same_week_day.and_hms(h, m, s)) // same week
                    } else {
                        Ok(same_week_day.and_hms(h, m, s) + Duration::days(7)) // next week
                    }
                }
            }
        }
        TimeClue::SameWeekDayAt(weekday, hms_maybe, am_or_pm_maybe) => {
            let (h, m, s) = hms_maybe.unwrap_or((0, 0, 0));
            let (h, m, s) = check_hms((h, m, s), am_or_pm_maybe)?;
            let monday = now.date() - Duration::days(now.weekday().num_days_from_monday() as i64);
            Ok((monday + Duration::days(weekday.num_days_from_monday() as i64)).and_hms(h, m, s))
        }
        TimeClue::ShortcutDayAt(rday, hms_maybe, am_or_pm_maybe) => {
            let (h, m, s) = hms_maybe.unwrap_or((0, 0, 0));
            let (h, m, s) = check_hms((h, m, s), am_or_pm_maybe)?;
            match rday {
                ShortcutDay::Today => Ok(now.date().and_hms(h, m, s)),
                ShortcutDay::Yesterday => Ok((now.date() - Duration::days(1)).and_hms(h, m, s)),
                ShortcutDay::Tomorrow => Ok((now.date() + Duration::days(1)).and_hms(h, m, s)),
            }
        }
        TimeClue::ISO((year, month, day), (h, m, s)) => {
            let utc = Utc.ymd_opt(year, month, day).and_hms_opt(h, m, s);
            match utc {
                LocalResult::Single(utc) => Ok(utc.with_timezone(&now.timezone())),
                _ => Err(EvaluationError::ChronoISOError {
                    year,
                    month,
                    day,
                    hour: h,
                    minute: m,
                    second: s,
                }),
            }
        }
    }
}

#[cfg(test)]
mod test {
    use crate::interpreter::{check_hms, evaluate, evaluate_time_clue};
    use crate::parser::AMPM::{AM, PM};
    use crate::parser::{Modifier, TimeClue};
    use chrono::offset::TimeZone;
    use chrono::Utc;
    use chrono::Weekday;

    #[test]
    fn test_check_hms() {
        assert_eq!(check_hms((19, 43, 42), None), Ok((19, 43, 42)));
        assert_eq!(check_hms((19, 43, 42), Some(AM)), Ok((19, 43, 42)));
        assert!(check_hms((19, 43, 42), Some(PM)).is_err());
        assert!(check_hms((24, 43, 42), None).is_err());
        assert!(check_hms((19, 63, 42), None).is_err());
        assert!(check_hms((19, 43, 62), None).is_err());
        assert_eq!(check_hms((6, 42, 43), Some(PM)), Ok((18, 42, 43)));
    }

    #[test]
    fn test_next_weekday() {
        let now = Utc
            .datetime_from_str("2020-07-12T12:45:00", "%Y-%m-%dT%H:%M:%S")
            .unwrap(); // sunday
        let expected = Utc
            .datetime_from_str("2020-07-17T00:00:00", "%Y-%m-%dT%H:%M:%S")
            .unwrap();
        assert_eq!(
            evaluate(
                TimeClue::RelativeDayAt(Modifier::Next, Weekday::Fri, None, None),
                now
            )
            .unwrap(),
            expected
        );
    }

    #[test]
    fn test_assume_next_day() {
        let now = Utc
            .datetime_from_str("2020-07-12T12:45:00", "%Y-%m-%dT%H:%M:%S")
            .unwrap(); // sunday

        let expected = Utc
            .datetime_from_str("2020-07-12T08:00:00", "%Y-%m-%dT%H:%M:%S")
            .unwrap();
        assert_eq!(
            evaluate_time_clue(TimeClue::Time((8, 0, 0), None), now.clone(), false).unwrap(),
            expected
        );

        let expected = Utc
            .datetime_from_str("2020-07-13T08:00:00", "%Y-%m-%dT%H:%M:%S")
            .unwrap();
        assert_eq!(
            evaluate_time_clue(TimeClue::Time((8, 0, 0), None), now, true).unwrap(),
            expected
        );
    }
}


src/parser.rs
use chrono::Weekday;
use pest::iterators::{Pair, Pairs};
use pest::Parser;
use std::fmt;
use std::fmt::Formatter;
use thiserror::Error;

#[derive(Parser)]
#[grammar = "time.pest"]
pub struct TimeParser;

pub type YMD = (i32, u32, u32);
pub type HMS = (u32, u32, u32);

#[derive(Error, Debug)]
pub enum ParseError {
    #[error("invalid integer")]
    ParseInt(#[from] std::num::ParseIntError),
    #[error(transparent)]
    PestError(#[from] pest::error::Error<Rule>),
    #[error("unexpected non matching pattern")]
    UnexpectedNonMatchingPattern,
    #[error("unknown weekday: `{0}`")]
    UnknownWeekday(String),
    #[error("unknown shortcut day: `{0}`")]
    UnknownShortcutDay(String),
    #[error("unknown modifier: `{0}`")]
    UnknownModifier(String),
    #[error("unknown quantifier `{0}`")]
    UnknownQuantifier(String),
    #[error("unknown am or pm `{0}`")]
    UnknownAMPM(String),
}

fn weekday_from(s: &str) -> Result<Weekday, ParseError> {
    match s {
        "monday" | "mon" => Ok(Weekday::Mon),
        "tuesday" | "tue" => Ok(Weekday::Tue),
        "wednesday" | "wed" => Ok(Weekday::Wed),
        "thursday" | "thu" => Ok(Weekday::Thu),
        "friday" | "fri" => Ok(Weekday::Fri),
        "saturday" | "sat" => Ok(Weekday::Sat),
        "sunday" | "sun" => Ok(Weekday::Sun),
        _ => Err(ParseError::UnknownWeekday(s.to_string())),
    }
}

#[derive(Error, Debug, PartialEq)]
pub enum AMPM {
    AM,
    PM,
}

impl fmt::Display for AMPM {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        match self {
            AMPM::AM => write!(f, "am"),
            AMPM::PM => write!(f, "pm"),
        }
    }
}

fn am_or_pm_from(s: &str) -> Result<AMPM, ParseError> {
    match s {
        "am" => Ok(AMPM::AM),
        "pm" => Ok(AMPM::PM),
        _ => Err(ParseError::UnknownAMPM(s.to_string())),
    }
}

#[derive(Debug, PartialEq)]
pub enum ShortcutDay {
    Today,
    Yesterday,
    Tomorrow,
}

fn shortcut_day_from(s: &str) -> Result<ShortcutDay, ParseError> {
    match s {
        "today" => Ok(ShortcutDay::Today),
        "yesterday" => Ok(ShortcutDay::Yesterday),
        "tomorrow" => Ok(ShortcutDay::Tomorrow),
        _ => Err(ParseError::UnknownShortcutDay(s.to_string())),
    }
}

#[derive(Debug, PartialEq)]
pub enum Modifier {
    Last,
    Next,
}

fn modifier_from(s: &str) -> Result<Modifier, ParseError> {
    match s {
        "last" => Ok(Modifier::Last),
        "next" => Ok(Modifier::Next),
        _ => Err(ParseError::UnknownModifier(s.to_string())),
    }
}

#[derive(Debug, PartialEq)]
pub enum Quantifier {
    Min,
    Hours,
    Days,
    Weeks,
    Months,
}

fn quantifier_from(s: &str) -> Result<Quantifier, ParseError> {
    match s {
        "min" => Ok(Quantifier::Min),
        "hours" | "hour" | "h" => Ok(Quantifier::Hours),
        "days" | "day" | "d" => Ok(Quantifier::Days),
        "weeks" | "week" | "w" => Ok(Quantifier::Weeks),
        "months" | "month" => Ok(Quantifier::Months),
        _ => Err(ParseError::UnknownQuantifier(s.to_string())),
    }
}

#[derive(Debug, PartialEq)]
pub enum TimeClue {
    /// Now.
    Now,
    /// Time without date: "19:43:42", "18", "8", "7pm", "3am"
    Time(HMS, Option<AMPM>),
    /// Relative (past) time clue: "4 minutes ago"
    Relative(usize, Quantifier),
    /// last/next `<day>` at `<time>`: "last friday at 12"
    RelativeDayAt(Modifier, Weekday, Option<HMS>, Option<AMPM>),
    /// Relative (future) time clue: "in 4 minutes"
    RelativeFuture(usize, Quantifier),
    /// Same week day at `<time>`: "monday at 4"
    SameWeekDayAt(Weekday, Option<HMS>, Option<AMPM>),
    /// `<shortcut_day>` at `<time>`: "yesterday at 4", "tomorrow"
    ShortcutDayAt(ShortcutDay, Option<HMS>, Option<AMPM>),
    /// YYYY-MM-DDThh:mm:ss or YYYY/MM/DDThh:mm:ss: "2020-12-25T19:43:00"
    ISO(YMD, HMS),
}

fn parse_time_hms(rules_and_str: &[(Rule, &str)]) -> Result<TimeClue, ParseError> {
    match rules_and_str {
        [(Rule::hms, h)] => {
            let h: u32 = h.parse()?;
            Ok(TimeClue::Time((h, 0, 0), None))
        }
        [(Rule::hms, h), (Rule::hms, m)] => {
            let h: u32 = h.parse()?;
            let m: u32 = m.parse()?;
            Ok(TimeClue::Time((h, m, 0), None))
        }
        [(Rule::hms, h), (Rule::hms, m), (Rule::hms, s)] => {
            let h: u32 = h.parse()?;
            let m: u32 = m.parse()?;
            let s: u32 = s.parse()?;
            Ok(TimeClue::Time((h, m, s), None))
        }
        [(Rule::hms, h), (Rule::am_or_pm, am_or_pm)] => {
            let h: u32 = h.parse()?;
            let am_or_pm = am_or_pm_from(am_or_pm)?;
            Ok(TimeClue::Time((h, 0, 0), Some(am_or_pm)))
        }
        [(Rule::hms, h), (Rule::hms, m), (Rule::am_or_pm, am_or_pm)] => {
            let h: u32 = h.parse()?;
            let m: u32 = m.parse()?;
            let am_or_pm = am_or_pm_from(am_or_pm)?;
            Ok(TimeClue::Time((h, m, 0), Some(am_or_pm)))
        }
        [(Rule::hms, h), (Rule::hms, m), (Rule::hms, s), (Rule::am_or_pm, am_or_pm)] => {
            let h: u32 = h.parse()?;
            let m: u32 = m.parse()?;
            let s: u32 = s.parse()?;
            let am_or_pm = am_or_pm_from(am_or_pm)?;
            Ok(TimeClue::Time((h, m, s), Some(am_or_pm)))
        }
        _ => Err(ParseError::UnexpectedNonMatchingPattern),
    }
}

fn parse_time_clue(pairs: &[Pair<Rule>]) -> Result<TimeClue, ParseError> {
    let rules_and_str: Vec<(Rule, &str)> = pairs
        .iter()
        .map(|pair| (pair.as_rule(), pair.as_str()))
        .collect();
    match rules_and_str.as_slice() {
        [(Rule::time_clue, _), (Rule::now, _), (Rule::EOI, _)] => Ok(TimeClue::Now),
        [(Rule::time_clue, _), (Rule::time, _), time_hms @ .., (Rule::EOI, _)] => {
            parse_time_hms(time_hms)
        }
        [(Rule::time_clue, _), (Rule::relative, _), (Rule::int, s), (Rule::quantifier, q), (Rule::EOI, _)] =>
        {
            let n: usize = s.parse()?;
            let q = quantifier_from(q)?;
            Ok(TimeClue::Relative(n, q))
        }
        [(Rule::time_clue, _), (Rule::relative_future, _), (Rule::int, s), (Rule::quantifier, q), (Rule::EOI, _)] =>
        {
            let n: usize = s.parse()?;
            let q = quantifier_from(q)?;
            Ok(TimeClue::RelativeFuture(n, q))
        }
        [(Rule::time_clue, _), (Rule::day_at, _), (Rule::mday, _), mday @ .., (Rule::EOI, _)] => {
            match mday {
                [(Rule::modifier, m), (Rule::weekday, w), (Rule::time, _), time_hms @ ..] => {
                    let (time_maybe, am_or_pm_maybe) = match parse_time_hms(time_hms)? {
                        TimeClue::Time(hms, am_or_pm) => (Some(hms), am_or_pm),
                        _ => (None, None),
                    };
                    let m = modifier_from(m)?;
                    let w = weekday_from(w)?;
                    Ok(TimeClue::RelativeDayAt(m, w, time_maybe, am_or_pm_maybe))
                }
                [(Rule::modifier, m), (Rule::weekday, w)] => {
                    let m = modifier_from(m)?;
                    let w = weekday_from(w)?;
                    Ok(TimeClue::RelativeDayAt(m, w, None, None))
                }
                [(Rule::weekday, w)] => {
                    let w = weekday_from(w)?;
                    Ok(TimeClue::SameWeekDayAt(w, None, None))
                }
                [(Rule::weekday, w), (Rule::time, _), time_hms @ ..] => {
                    let (time_maybe, am_or_pm_maybe) = match parse_time_hms(time_hms)? {
                        TimeClue::Time(hms, am_or_pm) => (Some(hms), am_or_pm),
                        _ => (None, None),
                    };
                    let w = weekday_from(w)?;
                    Ok(TimeClue::SameWeekDayAt(w, time_maybe, am_or_pm_maybe))
                }
                [(Rule::shortcut_day, r), (Rule::time, _), time_hms @ ..] => {
                    let (time_maybe, am_or_pm_maybe) = match parse_time_hms(time_hms)? {
                        TimeClue::Time(hms, am_or_pm) => (Some(hms), am_or_pm),
                        _ => (None, None),
                    };
                    let r = shortcut_day_from(r)?;
                    Ok(TimeClue::ShortcutDayAt(r, time_maybe, am_or_pm_maybe))
                }
                [(Rule::shortcut_day, r)] => {
                    let r = shortcut_day_from(r)?;
                    Ok(TimeClue::ShortcutDayAt(r, None, None))
                }
                _ => Err(ParseError::UnexpectedNonMatchingPattern),
            }
        }
        [(Rule::time_clue, _), (Rule::iso, _), (Rule::year, y), (Rule::month, m), (Rule::day, d), time_hms @ .., (Rule::EOI, _)] => {
            match parse_time_hms(time_hms)? {
                TimeClue::Time(hms, _) => {
                    let y: i32 = y.parse()?;
                    let m: u32 = m.parse()?;
                    let d: u32 = d.parse()?;
                    Ok(TimeClue::ISO((y, m, d), hms))
                }
                _ => Err(ParseError::UnexpectedNonMatchingPattern),
            }
        }
        [(Rule::time_clue, _), (Rule::date, _), (Rule::day, d), (Rule::month, m), (Rule::year, y), (Rule::EOI, _)] =>
        {
            let y: i32 = y.parse()?;
            let m: u32 = m.parse()?;
            let d: u32 = d.parse()?;
            Ok(TimeClue::ISO((y, m, d), (0, 0, 0)))
        }
        _ => Err(ParseError::UnexpectedNonMatchingPattern),
    }
}

/// Parse time clue from `s`. Prefer `htp::parse`.
///
/// This function is provided in case you wish to interpret time clues
/// yourself. Prefer `htp::parse`.
pub fn parse_time_clue_from_str(s: &str) -> Result<TimeClue, ParseError> {
    let pairs: Pairs<Rule> = TimeParser::parse(Rule::time_clue, s)?;
    let pairs: Vec<Pair<Rule>> = pairs.flatten().collect();
    parse_time_clue(pairs.as_slice())
}

#[cfg(test)]
mod test {
    use crate::parser::{
        parse_time_clue_from_str, Modifier, Quantifier, ShortcutDay, TimeClue, AMPM,
    };
    use chrono::Weekday;

    #[test]
    fn test_parse_time_ok() {
        assert_eq!(
            TimeClue::Time((9, 0, 0), None),
            parse_time_clue_from_str("9").unwrap()
        );
        assert_eq!(
            TimeClue::Time((9, 0, 0), Some(AMPM::AM)),
            parse_time_clue_from_str("9 am").unwrap()
        );
        assert_eq!(
            TimeClue::Time((9, 0, 0), Some(AMPM::PM)),
            parse_time_clue_from_str("9 pm").unwrap()
        );
        assert_eq!(
            TimeClue::Time((9, 30, 0), None),
            parse_time_clue_from_str("9:30").unwrap()
        );
        assert_eq!(
            TimeClue::Time((9, 30, 56), None),
            parse_time_clue_from_str("9:30:56").unwrap()
        );
    }

    #[test]
    fn test_parse_relative_ok() {
        for s in vec!["2 min ago", "2min ago", "2minago", "2   min  ago"].iter() {
            assert_eq!(
                TimeClue::Relative(2, Quantifier::Min),
                parse_time_clue_from_str(s).unwrap()
            );
        }
        for s in vec!["2 h ago", "2 hour ago", "2 hours ago"].iter() {
            assert_eq!(
                TimeClue::Relative(2, Quantifier::Hours),
                parse_time_clue_from_str(s).unwrap()
            );
        }
        for s in vec!["2 d ago", "2 day ago", "2 days ago"].iter() {
            assert_eq!(
                TimeClue::Relative(2, Quantifier::Days),
                parse_time_clue_from_str(s).unwrap()
            );
        }
    }

    #[test]
    fn test_parse_relative_future_ok() {
        for s in vec!["in 2 min", "in 2min", "in2min", "in  2   min"].iter() {
            assert_eq!(
                TimeClue::RelativeFuture(2, Quantifier::Min),
                parse_time_clue_from_str(s).unwrap()
            );
        }
        for s in vec!["in 2 h", "in 2 hour", "in 2 hours"].iter() {
            assert_eq!(
                TimeClue::RelativeFuture(2, Quantifier::Hours),
                parse_time_clue_from_str(s).unwrap()
            );
        }
        for s in vec!["in 2 d", "in 2 day", "in 2 days"].iter() {
            assert_eq!(
                TimeClue::RelativeFuture(2, Quantifier::Days),
                parse_time_clue_from_str(s).unwrap()
            );
        }
    }

    #[test]
    fn test_parse_shortcut_day_ok() {
        assert_eq!(
            TimeClue::ShortcutDayAt(ShortcutDay::Today, None, None),
            parse_time_clue_from_str("today").unwrap()
        );
        assert_eq!(
            TimeClue::ShortcutDayAt(ShortcutDay::Today, Some((7, 0, 0)), None),
            parse_time_clue_from_str("today at 7").unwrap()
        );
        assert_eq!(
            TimeClue::ShortcutDayAt(ShortcutDay::Yesterday, None, None),
            parse_time_clue_from_str("yesterday").unwrap()
        );
        assert_eq!(
            TimeClue::ShortcutDayAt(ShortcutDay::Yesterday, Some((19, 43, 0)), None),
            parse_time_clue_from_str("yesterday at 19:43").unwrap()
        );
        assert_eq!(
            TimeClue::ShortcutDayAt(ShortcutDay::Yesterday, Some((19, 43, 0)), None),
            parse_time_clue_from_str("yesterday at 19:43:00").unwrap()
        );
        assert_eq!(
            TimeClue::ShortcutDayAt(ShortcutDay::Tomorrow, Some((19, 43, 0)), None),
            parse_time_clue_from_str("tomorrow at 19:43:00").unwrap()
        );
    }

    #[test]
    fn test_parse_relative_day_ok() {
        assert_eq!(TimeClue::Now, parse_time_clue_from_str("now").unwrap());
        assert_eq!(
            TimeClue::SameWeekDayAt(Weekday::Fri, Some((19, 43, 0)), None),
            parse_time_clue_from_str("friday at 19:43").unwrap()
        );
        assert_eq!(
            TimeClue::RelativeDayAt(Modifier::Last, Weekday::Fri, None, None),
            parse_time_clue_from_str("last friday").unwrap()
        );
        assert_eq!(
            TimeClue::RelativeDayAt(Modifier::Last, Weekday::Fri, Some((9, 0, 0)), None),
            parse_time_clue_from_str("last friday at 9").unwrap()
        );
    }

    #[test]
    fn test_parse_same_week_ok() {
        let weekdays = vec![
            (Weekday::Mon, "monday"),
            (Weekday::Tue, "tuesday"),
            (Weekday::Wed, "wednesday"),
            (Weekday::Thu, "thursday"),
            (Weekday::Fri, "friday"),
            (Weekday::Sat, "saturday"),
            (Weekday::Sun, "sunday"),
        ];
        for (weekday, weekday_str) in weekdays.iter() {
            assert_eq!(
                TimeClue::SameWeekDayAt(weekday.clone(), None, None),
                parse_time_clue_from_str(weekday_str).unwrap()
            )
        }
        let weekdays = vec![
            (Weekday::Mon, "mon"),
            (Weekday::Tue, "tue"),
            (Weekday::Wed, "wed"),
            (Weekday::Thu, "thu"),
            (Weekday::Fri, "fri"),
            (Weekday::Sat, "sat"),
            (Weekday::Sun, "sun"),
        ];
        for (weekday, weekday_str) in weekdays.iter() {
            assert_eq!(
                TimeClue::SameWeekDayAt(weekday.clone(), None, None),
                parse_time_clue_from_str(weekday_str).unwrap()
            )
        }
        assert_eq!(
            TimeClue::SameWeekDayAt(Weekday::Fri, Some((19, 43, 0)), None),
            parse_time_clue_from_str("friday at 19:43").unwrap()
        );
    }

    #[test]
    fn test_parse_now_ok() {
        assert_eq!(TimeClue::Now, parse_time_clue_from_str("now").unwrap());
    }

    #[test]
    fn test_parse_iso_ok() {
        assert_eq!(
            TimeClue::ISO((2020, 12, 25), (19, 43, 42)),
            parse_time_clue_from_str("2020-12-25T19:43:42").unwrap()
        );

        assert_eq!(
            TimeClue::ISO((2020, 12, 25), (0, 0, 0)),
            parse_time_clue_from_str("25/12/2020").unwrap()
        );

        assert_eq!(
            TimeClue::ISO((2020, 12, 25), (0, 0, 0)),
            parse_time_clue_from_str("25-12-2020").unwrap()
        );
    }
}
